maro/simulator/scenarios/finance/business_engine.py

```diff
@@ -0,0 +1,653 @@
+
+        adjusted = True
+
+        while adjusted:
+            adjusted = False
+
+            executing_price = market_price
+
+            executing_price = slippage.execute(action.direction, executing_volume, executing_price, market_volume)
+
+            executing_commission = commission.execute(action.direction, executing_price, executing_volume)
+
+            executing_tax = tax.execute(action.direction, executing_price, executing_volume)
+
+            # constraint
+            if not self.validate_constrant(odd_shares, action.direction, executing_price, executing_volume):
+                executing_volume = self.adjust_constraint(
+                    odd_shares, action.direction, executing_price, executing_volume
+                )
+                adjusted = True
+                # print(f"Adjust volume for constraint:{executing_volume}")
+                continue
+
+            # money
+            if not self.validate_trade(
+                action.direction, executing_price, executing_volume,
+                executing_commission, executing_tax, remaining_money
+            ):
+                executing_volume = self.adjust_trade(
+                    action.direction, executing_price, executing_volume,
+                    executing_commission, executing_tax, remaining_money
+                )
+                adjusted = True
+                # print(f"Adjust volume for trade:{executing_volume}")
+                continue
+
+        if executing_volume > 0:
+            ret = TradeResult(executing_volume, executing_price, executing_commission, executing_tax)
+            rets.append(ret)
+        else:
+            pass
+
+        return rets
+
+    def validate_constrant(
+        self, odd_shares: int, direction: OrderDirection, executing_price: float, executing_volume: int
+    ):
+        ret = True
+        if direction == OrderDirection.buy:
+            if executing_volume % self._conf["trade_constraint"]["min_buy_unit"] != 0:
+                ret = False
+        if direction == OrderDirection.sell:
+            odd_volume = executing_volume % self._conf["trade_constraint"]["min_sell_unit"]
+            if odd_volume != 0:
+                if odd_shares != odd_volume:
+                    ret = False
+
+        return ret
+
+    def adjust_constraint(
+        self, odd_shares: int, direction: OrderDirection, executing_price: float, executing_volume: int
+    ):
+        ret = executing_volume
+        if direction == OrderDirection.buy:
+            odd_volume = ret % self._conf["trade_constraint"]["min_buy_unit"]
+            ret -= odd_volume
+        if direction == OrderDirection.sell:
+            odd_volume = executing_volume % self._conf["trade_constraint"]["min_sell_unit"]
+            if odd_volume != 0:
+                ret -= odd_volume
+                ret += odd_shares
+
+        return int(ret)
+
+    def validate_trade(
```

Comment:
rename ``validate_trade`` to ``validate_trade_on_remaining_money``?
rename ``adjust_trade`` to ``adjust_trade_on_remaining_money``?

Any other proposal?